topi/python/topi/cuda/nms.py

```diff
@@ -4,9 +4,184 @@
 import tvm
 
 from tvm import api
-from topi.vision import non_max_suppression
+from tvm.intrin import if_then_else
+from topi.vision import non_max_suppression, get_valid_counts
 from ..util import get_const_tuple
 
+
+def get_valid_counts_pre(data, flag, idx, score_threshold):
+    """Low level IR to get valid count of bounding boxes
+    given a score threshold. Also moves valid boxes to the
+    top of input data.
+
+    Parameters
+    ----------
+    data: Buffer
+        3D Buffer with shape [batch_size, num_anchors, 6], output of nms.
+
+    flag : Buffer
+        1D Buffer of flag indicating valid data with [num_anchors].
+
+    idx : Buffer
+        1D Buffer of valid data indices with [num_anchors].
+
+    score_threshold: float32
+        Lower limit of score for valid bounding boxes.
+
+    Returns
+    -------
+    stmt : Stmt
+        The result IR statement.
+    """
+    batch_size = data.shape[0]
+    num_anchors = data.shape[1]
+
+    ib = tvm.ir_builder.create()
+
+    data = ib.buffer_ptr(data)
+    flag = ib.buffer_ptr(flag)
+    idx = ib.buffer_ptr(idx)
+    score_threshold = tvm.make.node("FloatImm", dtype="float32", value=score_threshold)
+
+    max_threads = int(math.sqrt(
+        tvm.target.current_target(allow_none=False).max_num_threads))
+    nthread_tx = max_threads
+    nthread_bx = batch_size * num_anchors // max_threads + 1
+    tx = tvm.thread_axis("threadIdx.x")
+    bx = tvm.thread_axis("blockIdx.x")
+    ib.scope_attr(tx, "thread_extent", nthread_tx)
+    ib.scope_attr(bx, "thread_extent", nthread_bx)
+    tid = bx * max_threads + tx
+
+    with ib.if_scope(tid < batch_size * num_anchors):
+        i = tid / num_anchors # number of batches
+        j = tid % num_anchors # number of anchors
+        base_idx = i * num_anchors * 6
+        with ib.if_scope(data[base_idx + j * 6 + 1] > score_threshold):
+            flag[tid] = 1
+            idx[tid] = 1
+        with ib.else_scope():
+            flag[tid] = 0
+            idx[tid] = 0
+
+    with ib.if_scope(tid < batch_size):
+        with ib.for_range(0, num_anchors) as k:
+            with ib.if_scope(k > 0):
+                idx[tid * num_anchors + k] += idx[tid * num_anchors + k - 1]
+            ib.emit(tvm.make.Call(None, 'tvm_storage_sync',
+                                  tvm.convert(['shared']),
+                                  tvm.expr.Call.Intrinsic, None, 0))
+
+    return ib.get()
+
+def get_valid_counts_ir(data, flag, idx, valid_count, out):
```

Comment:
Is it possible to merge get_valid_counts_pre into get_valid_counts_ir?

Reply:
No, it's for global synchronization.